File: isabella.ml/src/canon/canon.ml
(** Canon: Formally verified lattice cryptography from Isabelle/HOL

    This module provides access to all verified functions from the Canon library.
    All code is extracted from proven-correct Isabelle specifications.

    Modules:
    - {!Zq} - Modular arithmetic over Z_q
    - {!Listvec} - Vector and matrix operations
    - {!Norms} - Vector norms and bounds
    - {!Decomp} - Gadget decomposition
    - {!Lwe_def} - LWE problem definition
    - {!Sis_def} - SIS problem definition
    - {!Polymod} - Polynomial ring arithmetic
    - {!Modulelwe} - Module-LWE and Module-SIS
    - {!Ntt} - Number Theoretic Transform (O(n log n) Cooley-Tukey)
    - {!Regev_pke} - Regev public-key encryption
    - {!Commit_sis} - SIS-based commitment scheme
    - {!Repeated_fs} - Domain-separated Fiat-Shamir transcript helpers
    - {!Confidential_sampling} - CSPRNG-backed confidential-transfer mask sampling
    - {!Confidential_balance} - Confidential balance proof helpers
    - {!Confidential_range} - Confidential range proof helpers
    - {!Confidential_transaction} - Confidential transaction proof helpers
    - {!Confidential_merkle} - Cryptographic confidential-note Merkle helpers
    - {!Kyber} - CRYSTALS-Kyber (ML-KEM) key encapsulation
    - {!Dilithium} - CRYSTALS-Dilithium (ML-DSA) digital signatures *)

module Zq = Zq
module Listvec = Listvec
module Norms = Norms
module Decomp = Decomp
module Lwe_def = Lwe_def
module Sis_def = Sis_def
module Polymod = Polymod
module Modulelwe = Modulelwe
module Ntt = Ntt
module Regev_pke = Regev_pke
module Commit_sis = Commit_sis
module Repeated_fs = Repeated_fs
module Confidential_sampling = Confidential_sampling
module Confidential_balance = Confidential_balance
module Confidential_range = Confidential_range
module Confidential_transaction = Confidential_transaction
module Confidential_merkle = Confidential_merkle
module Kyber = Kyber
module Dilithium = Dilithium

File: isabella.hs/src/Canon.hs
{-# LANGUAGE EmptyDataDecls, RankNTypes, ScopedTypeVariables #-}

-- | Canon: Formally verified lattice cryptography from Isabelle/HOL
--
-- This module re-exports all functionality from the Canon library.
-- All code is extracted from proven-correct Isabelle specifications.
--
-- = Overview
--
-- The Canon library provides the foundational building blocks for
-- lattice-based cryptography:
--
-- * "Canon.Algebra.Zq" - Modular arithmetic over Z_q
-- * "Canon.Linear.ListVec" - Vector and matrix operations
--
-- = Example Usage
--
-- @
-- import Canon
--
-- -- Centered modular reduction
-- x = mod_centered 7 5  -- Result: 2 (since 7 mod 5 = 2, and 2 <= 5/2)
--
-- -- Vector operations
-- v1 = [1, 2, 3]
-- v2 = [4, 5, 6]
-- dot = inner_prod v1 v2  -- Result: 32
--
-- -- Bit encoding for LWE
-- encoded = encode_bit 256 True   -- Result: 128 (q/2)
-- decoded = decode_bit 256 130    -- Result: True (close to q/2)
-- @
module Canon (
    -- * Re-exports
    module Canon.Algebra.Zq,
    module Canon.Linear.ListVec
) where

import Canon.Algebra.Zq
import Canon.Linear.ListVec
